libparsec/crates/client/src/workspace/store/per_manifest_update_lock.rs

```diff
@@ -0,0 +1,172 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use libparsec_platform_async::event::{Event, EventListener};
+use libparsec_types::prelude::*;
+
+/// This structure provide a way to lock independently each manifest for update.
+/// Once taken, a `ManifestUpdateLockGuard` is returned to the caller, which must
+/// be in turn provided to `release` once the lock should be released.
+#[derive(Debug)]
+pub(super) struct PerManifestUpdateLock {
+    per_manifest_lock: Vec<(VlobID, EntryLockState)>,
+}
+
+/// We cannot just use a simple async mutex to protect update of manifests.
+///
+/// This is because the manifests cache is itself behind a sync mutex (which
+/// must be released before waiting on the per manifest async mutex !).
+/// The natural solution to this is to wrap the per manifest async mutex in an
+/// Arc, but this creates a new issue: the `ManifestUpdateLockGuard` would then
+/// have to keep both the `Arc<AsyncMutex>` and the corresponding `AsyncMutexGuard`.
+/// However this is not possible given the guard takes a reference on the mutex !
+///
+/// Hence we basically have to re-implement a custom async mutex using the lower-level
+/// event system.
+#[derive(Debug)]
+enum EntryLockState {
+    /// A single coroutine has locked the manifest
+    Taken,
+    /// Multiple coroutines want to lock the manifest: the first has the lock
+    /// and the subsequent ones are listening on the event (which will be fired
+    /// when the first coroutine releases the manifest).
+    /// Note this means all coroutines waiting for the event will fight with no
+    /// fairness to take the lock (it's basically up to the tokio scheduler).
+    TakenWithConcurrency(Event),
+}
+
+pub(super) enum ManifestUpdateLockTakeOutcome {
+    // The updater object is responsible to release the update lock on drop.
+    // Hence this object must be created as soon as possible otherwise a deadlock
+    // could occur (typically in case of an early return due to error handling)
+    Taken(ManifestUpdateLockGuard),
+    NeedWait(EventListener),
+}
+
+/// ⚠️ This lock doesn't release on drop ⚠️
+///
+/// Instead you must manually call `PerManifestUpdateLock::release` once you are done.
+#[derive(Debug)] // Note this should not be made `Clone` !
+pub(super) struct ManifestUpdateLockGuard {
+    entry_id: VlobID,
+    #[allow(unused)]
```

Comment:
This field seem to be used below.

Reply:
This `is_released` boolean has only one purpose: allowing the `Drop` to ensure the the guard has been released (the release cannot be done in the `Drop` given it involves having access to a specific third party object)

However the `Drop` code has been commented for the moment: the issue anytime an arbitrary panic occurs, the guard is not released, so the `Drop` in turn panic and shadow the initial panic !

So for the moment this is disabled... but we must do something about it !
I've created any issue about that: https://github.com/Scille/parsec-cloud/issues/7367